### ISSUE TITLE
Meshes ending with "_Wheel" are incorrectly assigned a VehicleWheel3D node

### ISSUE THREAD

**Issue by allenwp** [godotengine/godot#72419]
### Godot version

v3.5.1.stable.official [6fed1ffa3] and v4.0.beta16.official [518b9e580]

### System information

Windows 11, GLES3 or Forward+

### Issue description

Any time a mesh inside of a 3D model, such as a glTF 2.0 file, ends with the string "_Wheel", Godot will insert a VehicleWheel3D node as a parent of the mesh and will remove "_Wheel" from the mesh name. This results in at least two problems I can think of:

1. The mesh is a steering wheel named "Steering_Wheel". In this case, this is not a vehicle wheel and should not have the VehicleWheel3D node associated with it. (Note: This is the specific situation that happened with my project)
2. The project may not be designed to have wheels associated with vehicles. So any time a "_Wheel" comes up in the model, it is incorrectly assigned a VehicleWheel3D when there will never be the required VehicleBody node as a parent.

![image](https://user-images.githubusercontent.com/17506573/215610298-6dfb2f9a-bb76-4253-aec1-b93445efef9a.png)

Because there is specific configuration required to make a VehicleWheel3D node function correctly (a VehicleBody node must be a parent), I believe that it would be better to not assign the VehicleWheel3D node to meshes upon import, but instead require the user to add that node when they actively choose to, such as when they are configuring the other vehicle nodes like the VihicleBody.

### Steps to reproduce

1. Import a model with a mesh that is named something that ends with "_Wheel"
2. Note that a VehicleWheel3D is added to the scene tree of the model

### Minimal reproduction project

[Godot3_Wheel_Bug.zip](https://github.com/godotengine/godot/files/10540529/Godot3_Wheel_Bug.zip)

**Comment by AThousandShips** [comment]
> when there will never be the required VehicleBody node as a parent

With this do you mean for situations where you haven't made a VehicleBody in the file, or in general?

**Comment by allenwp** [comment]
> > when there will never be the required VehicleBody node as a parent
> 
> With this do you mean for situations where you haven't made a VehicleBody in the file, or in general?

I was referring to situations where there never will be a VehicleBody in general for that project.

**Comment by allenwp** [comment]
Upon further reflection, I mean to bring up the issue of reserved mesh name patterns in general.

I think that these sorts of reserved name patters are generally problematic because they are likely to be unknown to many or all members of a game development team. That these name patterns may be unknown to the team members, and even more likely to the 3D artists of the team, can result in wasteful team member interactions like this example:

Programmer to 3D artist: “About that model you just sent me: It’s great and you’ve done a really good job with it… but can you re-export it for me with the ‘Steering_Wheel’ mesh renamed to something like ‘Steering_Wheel_’? The specific mesh name you used is giving me issues with the game engine.”

When reserved name patterns are used in the import stage for a 3D model’s meshes, it results in more back-and-forth between artists and programmers, which slows down the project. Having the programmer simply add in the correct nodes after import is a more streamlined approach when multiple team members are involved or when it may be non-trivial to re-export the 3D model.

As a possible solution to this problem, I would suggest moving this sort of mesh name matching behaviour outside of the core engine and into an extension. This way, the core import code is simplified and developers who want this specific name matching behaviour for their game project can still get it with the extension. I believe that mesh-name-specific import behaviour should be a proactive choice of the development team, rather than a default.

**Comment by AThousandShips** [comment]
Moving this out of core would be a big change, but making it optional and configurable would be an option, the convenience of the functionality isn't to be ignored and the benefit of having it in core, as it already exists

**Comment by aaronfranke** [comment]
> @allenwp: I think that these sorts of reserved name patters are generally problematic [...]

I completely agree, and this is why on the glTF side, we are working on specifying more of this information explicitly in the glTF file, with standards such as [OMI_physics_body](https://github.com/omigroup/gltf-extensions/tree/main/extensions/2.0/OMI_physics_body) and [OMI_vehicle_wheel](https://github.com/aaronfranke/gltf-extensions/tree/OMI_vehicle/extensions/2.0/OMI_vehicle_wheel).

With these standards, instead of naming a node "MyTire_Wheel" to import it as "MyTire" with type "VehicleWheel3D", you will be able to name a node "MyTire" and give it metadata of `"extensions": { "OMI_vehicle_wheel": { "wheel": 0 } }` to explicitly mark it as a wheel. Further settings can then be specified in the wheel settings array at index 0. The GUI for specifying this metadata will need to be implemented in Blender, likely as a plugin. Or, if you don't want to install such a plugin in Blender, you could configure it in Godot instead.

The existing system of using node name suffixes is something I consider to be a hack, and I will be happy when we can remove this hack from the engine. But this cannot be done until the glTF extensions that replace it are more mature, including support in Blender. It also cannot be done until the next compatibility breakage. As for your suggestion of moving it to an extension, I'd say no, we'll keep it until it's not needed and then it will be gone.

Another point, the current system for node name suffixes is done in Godot's general scene import code, not in the code of any particular 3D model format. This means that it works for glTF, FBX, Collada, etc, automatically. However, the node name suffix system is only supported in the editor, not when importing files at runtime. This system will also break any files that depend on the original node structure, such as glTF animations pointing to a node named "MyTire_Wheel" that cannot handle the node being renamed to "MyTire" after the glTF code finishes.

**Comment by akien-mga** [comment]
> However, the node name suffix system is only supported in the editor, not when importing files at runtime. This system will also break any files that depend on the original node structure, such as glTF animations pointing to a node named "MyTire_Wheel" that cannot handle the node being renamed to "MyTire" after the glTF code finishes.

I'd say this is a bug, and not a reason to deprecate/remove that system. I don't see what would prevent handling these hints at runtime import too?

**Comment by aaronfranke** [comment]
@akien-mga It's a consequence of how this was designed. The code for the suffixes is handled in the editor-specific ResourceImporterScene class, which is not used when importing a file with GLTFDocument at runtime.

ResourceImporterScene is designed to allow configuring how a 3D model is imported as a Godot scene, so most of it is not useful for runtime imports, where you don't know what the file will contain in advance, and therefore the solution can't be to allow using this class at runtime. We also don't want to move the code into the individual formats, since then it would have to be repeated for all of them, and would not align with the goals of glTF physics. The only reasonable solution I could think of would be to effectively split ResourceImporterScene into two pieces (non-equally-sized), providing only parts useful for runtime imports in one class available at runtime, and it's not a refactor I want to do if my goal is to remove this feature in the coming decades.

**Comment by allenwp** [comment]
Sounds like a lot of these ideas can exist in a new proposal.

My original issue can probably be closed when all cases of node type assignments correctly respect scene requirements, such as only assigning `VehicleWheel3D` when a parent `VehicleBody` is also being assigned.
